controllers/metal3.io/hostfirmwaresettings_controller.go

```diff
@@ -0,0 +1,460 @@
+	if len(errors) == 0 {
+		info.hfs.Status.IsValid = true
+	} else {
+		for _, error := range errors {
+			info.publishEvent("ValidationFailed", fmt.Sprintf("Invalid BIOS setting: %v", error))
+		}
+		info.hfs.Status.IsValid = false
+	}
+
+	t := metav1.Now()
+	info.hfs.Status.LastUpdated = &t
+
+	return r.Status().Update(context.TODO(), info.hfs)
+}
+
+// Get a firmware schema that matches the host vendor or create one if it doesn't exist
+func (r *HostFirmwareSettingsReconciler) getOrCreateFirmwareSchema(info *rInfo, schema map[string]metal3v1alpha1.SettingSchema) (fSchema *metal3v1alpha1.FirmwareSchema, err error) {
+
+	info.log.Info("getting firmwareSchema")
+
+	schemaName := GetSchemaName(schema)
+	firmwareSchema := &metal3v1alpha1.FirmwareSchema{}
+
+	// If a schema exists that matches, use that, otherwise create a new one
+	if err = r.Get(context.TODO(), client.ObjectKey{Namespace: info.hfs.ObjectMeta.Namespace, Name: schemaName},
+		firmwareSchema); err == nil {
+
+		info.log.Info("found existing firmwareSchema resource")
+
+		return firmwareSchema, nil
+	}
+	if !k8serrors.IsNotFound(err) {
+		// Error reading the object
+		return nil, err
+
+	}
+
+	firmwareSchema = &metal3v1alpha1.FirmwareSchema{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      schemaName,
+			Namespace: info.hfs.ObjectMeta.Namespace,
+		},
+	}
+
+	// If available, store hardware details in schema for additional info
+	if info.bmh.Status.HardwareDetails != nil {
+		firmwareSchema.Spec.HardwareVendor = info.bmh.Status.HardwareDetails.SystemVendor.Manufacturer
+		firmwareSchema.Spec.HardwareModel = info.bmh.Status.HardwareDetails.SystemVendor.ProductName
+	}
+
+	// Copy in the schema from provisioner
+	firmwareSchema.Spec.Schema = make(map[string]metal3v1alpha1.SettingSchema)
+	for k, v := range schema {
+		firmwareSchema.Spec.Schema[k] = v
+	}
+
+	if err = r.Create(context.TODO(), firmwareSchema); err != nil {
+		return nil, err
+	}
+	// Set hfs as owner after the create
+	if controllerutil.SetOwnerReference(info.hfs, firmwareSchema, r.Scheme()); err != nil {
+		return nil, errors.Wrap(err, "could not set owner of firmwareSchema")
+	}
+	if err = r.Update(context.TODO(), firmwareSchema); err != nil {
+		return nil, err
+	}
+
+	info.log.Info("created new firmwareSchema resource")
+
+	return firmwareSchema, nil
+}
+
+// Create a hostFirmwareSettings
+func (r *HostFirmwareSettingsReconciler) newHostFirmwareSettings(info *rInfo, namespacedName types.NamespacedName) (err error) {
+
+	info.hfs.ObjectMeta = metav1.ObjectMeta{
+		Name: namespacedName.Name, Namespace: namespacedName.Namespace}
+	info.hfs.Status.Settings = make(metal3v1alpha1.SettingsMap)
+	info.hfs.Spec.Settings = make(metal3v1alpha1.DesiredSettingsMap)
+
+	if err = r.Create(context.TODO(), info.hfs); err != nil {
+		return errors.Wrap(err, "failure creating hostFirmwareSettings resource")
+	}
+	info.log.Info("created new hostFirmwareSettings resource")
+
+	return nil
```

Comment:
Yeah, good point, I can see that if the actual setting gets updated outside of Ironic it can cause a an inadvertent update to the old value. I've removed the syncing of the Spec on update. This will add an extra step for the user to create the entry in the Spec when making a change but it will prevent this spurious update.